File: crap4py/scripts/crapkit/report.py
from decimal import ROUND_HALF_UP, Decimal

from crapkit.score import CRAPPY_THRESHOLD, Entry, sort_entries

HEADER = f"{'Function':<30} {'Module':<35} {'CC':>4} {'Cov%':>7} {'CRAP':>8}"
ONE_DECIMAL = Decimal("0.1")


def round_half_up(value: float) -> float:
    return float(Decimal(str(value)).quantize(ONE_DECIMAL, rounding=ROUND_HALF_UP))


def format_table(entries: list[Entry]) -> str:
    lines = ["CRAP Report", "===========", HEADER, "-" * len(HEADER)]
    for entry in sort_entries(entries):
        lines.append(
            f"{entry.name:<30} {entry.module:<35} {entry.complexity:>4} "
            f"{_format_coverage(entry.coverage):>7} {_format_score(entry.crap):>8}"
        )
    lines.append("")
    return "\n".join(lines)


def _format_coverage(coverage: float | None) -> str:
    return "N/A" if coverage is None else f"{round_half_up(coverage):.1f}%"


def _format_score(score: float | None) -> str:
    return "N/A" if score is None else f"{round_half_up(score):.1f}"


def build_document(
    entries: list[Entry],
    language: str,
    paths: list[str],
    diff_base: str | None,
    generated_at: str,
) -> dict:
    ordered = sort_entries(entries)
    scores = [entry.crap for entry in ordered if entry.crap is not None]
    return {
        "version": 1,
        "language": language,
        "generated_at": generated_at,
        "filters": {"paths": paths, "diff_base": diff_base},
        "summary": {
            "functions": len(ordered),
            "crappy": sum(1 for score in scores if score >= CRAPPY_THRESHOLD),
            "max_crap": round_half_up(max(scores)) if scores else None,
        },
        "entries": [_serialize(entry) for entry in ordered],
    }


def _serialize(entry: Entry) -> dict:
    return {
        "name": entry.name,
        "module": entry.module,
        "file": entry.file,
        "start_line": entry.start_line,
        "end_line": entry.end_line,
        "complexity": entry.complexity,
        "coverage": None if entry.coverage is None else round_half_up(entry.coverage),
        "crap": None if entry.crap is None else round_half_up(entry.crap),
        "statements": entry.statements,
        "covered_statements": entry.covered_statements,
    }

File: crap4ts/scripts/lib/report.mjs
import { CRAPPY_THRESHOLD, crapScore, sortEntries } from "./score.mjs";

const padRight = (text, width) => String(text).padEnd(width);
const padLeft = (text, width) => String(text).padStart(width);

const HEADER =
  `${padRight("Function", 30)} ${padRight("Module", 35)} ` +
  `${padLeft("CC", 4)} ${padLeft("Cov%", 7)} ${padLeft("CRAP", 8)}`;

const formatCoverage = (coverage) =>
  coverage === null ? "N/A" : `${coverage.toFixed(1)}%`;

const formatScore = (score) => (score === null ? "N/A" : score.toFixed(1));

export function formatTable(entries) {
  const lines = ["CRAP Report", "===========", HEADER, "-".repeat(HEADER.length)];
  for (const entry of sortEntries(entries)) {
    const score = crapScore(entry.complexity, entry.coverage);
    lines.push(
      `${padRight(entry.name, 30)} ${padRight(entry.module, 35)} ` +
        `${padLeft(entry.complexity, 4)} ${padLeft(formatCoverage(entry.coverage), 7)} ` +
        `${padLeft(formatScore(score), 8)}`,
    );
  }
  lines.push("");
  return lines.join("\n");
}

const round = (value) => (value === null ? null : Math.round(value * 10) / 10);

export function buildDocument({ entries, language, paths, diffBase, generatedAt }) {
  const ordered = sortEntries(entries);
  const scores = ordered
    .map((entry) => crapScore(entry.complexity, entry.coverage))
    .filter((score) => score !== null);
  return {
    version: 1,
    language,
    generated_at: generatedAt,
    filters: { paths, diff_base: diffBase },
    summary: {
      functions: ordered.length,
      crappy: scores.filter((score) => score >= CRAPPY_THRESHOLD).length,
      max_crap: scores.length ? round(Math.max(...scores)) : null,
    },
    entries: ordered.map((entry) => ({
      name: entry.name,
      module: entry.module,
      file: entry.file,
      start_line: entry.startLine,
      end_line: entry.endLine,
      complexity: entry.complexity,
      coverage: round(entry.coverage),
      crap: round(crapScore(entry.complexity, entry.coverage)),
      statements: entry.statements,
      covered_statements: entry.coveredStatements,
    })),
  };
}
